### ISSUE TITLE
add create_trainer & create_evaluator util function

### ISSUE THREAD

**Issue by ipod825** [pytorch/ignite#446]
It would be nice to have two functions that wraps up creating trainer/evaluator using `Engine`. In case where the data batch is not in `(input, label)` form, there's still some boilerplate code as described [here](https://pytorch.org/ignite/quickstart.html#explanation). 
```python
def _update(engine, batch):
    model.train()
    optimizer.zero_grad()
    x, y = _prepare_batch(batch, device=device)
    y_pred = model(x)
    loss = loss_fn(y_pred, y)
    loss.backward()
    optimizer.step()
    return loss.item()
```
In such case, I think the following two help function would be very useful that the user only need to define a function for the `x, y = _prepare_batch(batch, device=device)` and `y_pred = model(x)` line:

```python
def create_trainer(model, optimizer, step_fn, device=None):
    """Factory function for creating a trainer for supervised models.

    Args:
        model (`torch.nn.Module`): the model to train
        optimizer (`torch.optim.Optimizer`): the optimizer to use
        step_fn (Callable): function that receives `model`, `batch`, `device` and outputs
            1. loss (Tensor): loss for backpropagation
            2. endpoints (dict: str -> Tensor): endpoints assigned to engine.state.output
        device (str, optional): device type specification (default: None).
            Applies to the model and passed to step_fn.

    Returns:
        Engine: a trainer engine with supervised update function
    """
    if device:
        model.to(device)

    def _update(engine, batch):
        model.train()
        optimizer.zero_grad()
        loss, endpoints = step_fn(model, batch, device=device)
        loss.backward()
        optimizer.step()
        return endpoints

    return Engine(_update)


def create_evaluator(model, step_fn, metrics={}, device=None):
    """Factory function for creating an evaluator for supervised models.

    Args:
        model (`torch.nn.Module`): the model to train
        step_fn (Callable): function that receives `model`, `batch`, `device` and outputs
            endpoints (dict: str -> Tensor): endpoints assigned to engine.state.output
        device (str, optional): device type specification (default: None).
            Applies to both model and batches.

    Returns:
        Engine: an evaluator engine with supervised inference function
    """
    if device:
        model.to(device)

    def _inference(engine, batch):
        model.eval()
        with torch.no_grad():
            endpoints = step_fn(model, batch, device=device)
            return endpoints

    engine = Engine(_inference)

    return engine

```

**Comment by vfdev-5** [comment]
@ipod825 there are many ways to handle specific cases different of `(inputs, labels)`. The functions like `create_supervised_trainer` or `create_supervised_evaluator` are just helper functions to create faster an Engine for a basic typical task with `(x, y)`. 

For other cases, we suggest to create directly the trainer and evaluator using `Engine` class:
```python

def processing_function(engine, batch):
     # do whatever you would like

trainer = Engine(processing_function)
```

In some cases, when your input data can be organized into two tuples `(x, y)` where, for example, `x = (img_512, img_224, other_scalar_data, ...)` and `y=(mask, bboxes, labels)` you can still work with `create_supervised_trainer` if your model accepts `x` correctly and produces predictions `y_pred` such that the criterion does not fail when called as `criterion(y_pred, y)`.

**Comment by ipod825** [comment]
I understand that I can create the trainer and evaluator using `Engine` class:
```python
def processing_function(engine, batch):
     # do whatever you would like

trainer = Engine(processing_function)
```

My point is that `processing_function(engine, batch)` will always be like:
```python

if device:
        model.to(device)
def processing_function(engine, batch):
        model.train()
        optimizer.zero_grad()

        # This two lines might be model/machine dependent
        batch = batch.to(device)
        loss, endpoints = model(model, batch, device=device)

        loss.backward()
        optimizer.step()
        return loss, endpoints
```

So I propose to free user from writing the boilerplate code and just write a `step_fn` for the two lines that might be model/machine dependent. In fact, `create_supervised_trainer` is just a special case with `step_fn` as:
```
def step_fn(engine, batch):
        batch = batch.to(device)
        y_pred = model(x)
        loss = loss_fn(y_pred, y)
        return loss, {}  # No endpoints is returned
```

**Comment by vfdev-5** [comment]
@ipod825 what are the endpoints exactly in you case ? 
In the examples, you mention, `step_fn` for trainer is different of `step_fn` for the evaluator. IMO, mixing inside the `step_fn` the forward pass and loss computation does not seems good for me...

**Comment by ipod825** [comment]
Note that in my first post, `_update` returns only `endpoints`, which then become engine.state.output.

`endpoints` contains intermediate tensors when computing the `loss` tensor. Binding them to engine.state.output is useful for defining callbacks. For example, you might want to save the generated fake sample in gan every epoch. Or for example, I am implementing Variational AutoEncoder and want to plot KL divergence and reconstruction loss separately using tensorboardX, I write:
```python
    writer = SummaryWriter(log_dir=log_dir)

    @trainer.on(Events.ITERATION_COMPLETED)
    def plot_train(engine):
        if engine.state.iteration % plot_interval == 0:
            writer.add_scalar("training/KL Divergence",
                              engine.state.output['kl_loss'].item(),
                              engine.state.iteration)
            writer.add_scalar("training/Reconstruction Loss",
                              engine.state.output['rec_loss'].item(),
                              engine.state.iteration)
```
where my `step_fn` is as the following:
```python
def train_step(model, batch, device):
    img, label = batch
    img = img.to(device, non_blocking=True)
    img_rec, mu, logvar = model(img)

    BCE = F.binary_cross_entropy(img_rec, img.view(-1, 784), reduction='sum')
    KLD = -0.5 * torch.sum(1 + logvar - mu.pow(2) - logvar.exp())
    loss = BCE + KLD
    return loss, {'rec_loss': BCE, 'kl_loss': KLD}

```

**Comment by ipod825** [comment]
Note that in my `create_evaluator`, I used `with torch.no_grad()`. So no gradient is not computed in testing phase. `step_fn` just defines the forward pass.

**Comment by vfdev-5** [comment]
@ipod825 Okay, I see better what you would like to do. Please, take a look at [this](https://pytorch.org/ignite/concepts.html#state) section. Especially, this part:
```python
def update(engine, batch):
    x, y = batch
    y_pred = model(inputs)
    loss = loss_fn(y_pred, y) # here you can output all your loss values
    optimizer.zero_grad()
    loss.backward()
    optimizer.step()
    return {'loss': loss.item(),  # here, similarly you can output them to `engine.state.output`
            'y_pred': y_pred,
            'y': y}

trainer = Engine(update)

@trainer.on(Events.EPOCH_COMPLETED)
def print_loss(engine):
    epoch = engine.state.epoch
    loss = engine.state.output['loss']
    print ('Epoch {epoch}: train_loss = {loss}'.format(epoch=epoch, loss=loss))

accuracy = Accuracy(output_transform=lambda x: [x['y_pred'], x['y']])
accuracy.attach(trainer, 'acc')
trainer.run(data, max_epochs=10)
```
and the note on the `State` and `new_attribute`. Probably, this would help you to solve your problems without changing the API of Ignite.

Anyway, we can think about on how to get your suggestions more generic than just to output other losses and if there is something interesting, these helpers can be implemented in `contrib` module.

**Comment by ipod825** [comment]
I see what I was missing. Definitely my VAE case can be just handled by  `create_supervised_trainer`. I was thinking about GAN that doesn't fit into this case. And I was trying to make a generic trainer by first writing the supervised_trainer in this manner. But after looking back to the GAN ignite example, I realize that my trainer can't actually handle such case. I'll give it more thought to see if I can come up with any new idea.

**Comment by ipod825** [comment]
I think using `def processing_function(engine, batch):` is the only possible way to deal with arbitrary training procedure. Sorry for the spammer.

**Comment by vfdev-5** [comment]
@ipod825 no problem ! Anyway, thanks for the feedback while using Ignite. 
About VAE, we have not yet merged a notebook on [this](https://github.com/pytorch/ignite/pull/415), maybe this could help ...
